libs/helpers/helpers.go

```diff
@@ -10,51 +11,53 @@ import (
 // LatestVersion gets a slice of versions and returns the latest version as a string
 // e.g. ["0.1", "0.10", "1.0"] or ["0.1.0", "0.10.1", "1.0.2"]
 // any number of minor/patch versions will work as long as all versions are consistent
-// i.e. this is not allowed ["0.1", "0.10.0", "1.0"] and the function will return ""
+// i.e. this is not allowed ["0.1", "0.10.0", "1.0"] and the function will return error
 func LatestVersion(vl []string) (string, error) {
 	if len(vl) == 0 {
 		return "", errors.New("Unable to determine latest version, empty slice")
 	}
 
-	var ss [][]string
-	var vlen int
+	// initialize latest with zeros
+	latest := []int{}
+	for i := 0; i < len(strings.Split(vl[0], ".")); i++ {
+		latest = append(latest, 0)
+	}
+
+	// loop over list and compare each to latest
+	for _, sv := range vl {
+		current := strings.Split(sv, ".")
+		if len(current) < 2 {
+			log.Printf("Error: format mismatch for version '%s'", sv)
+			return "", errors.New("Unable to determine latest version, format mismatch")
+		}
 
-	// split all versions into their atomic parts (e.g. "1.0" becomes ["1", "0"]) and put into slice of slices ss
-	for _, v := range vl {
-		sv := strings.Split(v, ".")
-		if vlen == 0 {
-			vlen = len(sv)
-		} else {
-			if vlen != len(sv) {
-				log.Printf("Error: format mismatch for version '%s'", v)
+		si := []int{}
+		for _, s := range current {
+			i, err := strconv.Atoi(s)
+			if err != nil {
+				log.Printf("Error: format mismatch for version '%s'", sv)
 				return "", errors.New("Unable to determine latest version, format mismatch")
 			}
+			si = append(si, i)
 		}
-		ss = append(ss, sv)
-	}
 
-	// bubble sort the slices starting with the minor version (the one on the right) and moving to the left
-	for l := vlen - 1; l >= 0; l-- {
-		for m := 0; m < len(ss); m++ {
-			for n := 0; n < len(ss)-1; n++ {
-				int1, err := strconv.Atoi(ss[n][l])
-				if err != nil {
-					log.Printf("Error: format mismatch for version '%s'", strings.Join(ss[n], "."))
-					return "", errors.New("Unable to determine latest version, format mismatch")
-				}
-				int2, err := strconv.Atoi(ss[n+1][l])
-				if err != nil {
-					log.Printf("Error: format mismatch for version '%s'", strings.Join(ss[n+1], "."))
-					return "", errors.New("Unable to determine latest version, format mismatch")
-				}
-				if int1 > int2 {
-					ss[n], ss[n+1] = ss[n+1], ss[n]
-				}
+		if len(latest) != len(si) {
+			log.Printf("Error: format mismatch for version '%s'", sv)
+			return "", errors.New("Unable to determine latest version, format mismatch")
+		}
+
+		for i, siv := range si {
+			// we break out unless the version index is equal to the latest one
+			if siv > latest[i] {
+				latest = si
+				break
+			}
+			if siv < latest[i] {
+				break
 			}
 		}
 	}
-
-	return strings.Join(ss[len(ss)-1], "."), nil
+	return strings.Trim(strings.Join(strings.Fields(fmt.Sprint(latest)), "."), "[]"), nil
```

Comment:
a lot of slow functions chained together here... even a straight append instead is a lot better

```golang
package main

import(
	"fmt"
	"strings"
	"strconv"
	"testing"
)

func IntSliceString1(nums []int, separator string) string {
	return strings.Trim(strings.Join(strings.Fields(fmt.Sprint(nums)), separator), "[]")
}

func IntSliceString2(nums []int, separator string) string {
	if len(nums) == 0 {
		return ""
	}

	s := make([]string, len(nums))
	for i, n := range nums {
		s[i] = strconv.Itoa(n)
	}

	return strings.Join(s, separator)
}

func IntSliceString3(nums []int, separator string) string {
	if len(nums) == 0 {
		return ""
	}

	s := []string{}
	for _, n := range nums {
		s = append(s, strconv.Itoa(n))
	}

	return strings.Join(s, separator)
}

func Benchmark1(b *testing.B) {
	foo := []int{1,2,3,4,5,6,7,8,9,10,100,1000}
	for n := 0; n < b.N; n++ {
		IntSliceString1(foo, ".")
	}
}

func Benchmark2(b *testing.B) {
	foo := []int{1,2,3,4,5,6,7,8,9,10,100,1000}
	for n := 0; n < b.N; n++ {
		IntSliceString2(foo, ".")
	}
}

func Benchmark3(b *testing.B) {
	foo := []int{1,2,3,4,5,6,7,8,9,10,100,1000}
	for n := 0; n < b.N; n++ {
		IntSliceString3(foo, ".")
	}
}
```

```
% ➜  tmp go test -bench=. -benchtime=20s
goos: darwin
goarch: amd64
Benchmark1-8   	20000000	      1950 ns/op
Benchmark2-8   	100000000	       417 ns/op
Benchmark3-8   	50000000	       727 ns/op
PASS
ok  	_/Users/fish/tmp/go/yale/tmp	120.277s
```

Reply:
but it's a one-liner 😉

Reply:
the benchmark stuff is pretty cool!